src/Control/Effect/Carrier.hs

```diff
@@ -21,6 +21,13 @@ class HFunctor h where
   -- | Higher-order functor map of a natural transformation over higher-order positions within the effect.
   hmap :: (forall x . m x -> n x) -> (h m a -> h n a)
 
+  -- | When an 'HFunctor' contains no recursive occurrences of @m@, 'hmap' can default to 'coerce'.
+  default hmap :: Coercible (h m a) (h n a)
+               => (forall x . m x -> n x)
+               -> (h m a -> h n a)
+  hmap _ = coerce
+  {-# INLINE hmap #-}
```

Comment:
What do the error messages look like when you try to use the default definition with a higher-order effect?

Reply:
It could be worse, but it could be better:
```
src/Control/Effect/Writer.hs:28:1: error:
    • Couldn't match representation of type ‘m’ with that of ‘n’
        arising from a use of ‘Control.Effect.Carrier.$dmhmap’
      ‘m’ is a rigid type variable bound by
        the type signature for:
          hmap :: forall (m :: * -> *) (n :: * -> *) a.
                  (forall x. m x -> n x) -> Writer w m a -> Writer w n a
        at src/Control/Effect/Writer.hs:28:1-37
      ‘n’ is a rigid type variable bound by
        the type signature for:
          hmap :: forall (m :: * -> *) (n :: * -> *) a.
                  (forall x. m x -> n x) -> Writer w m a -> Writer w n a
        at src/Control/Effect/Writer.hs:28:1-37
    • In the expression: Control.Effect.Carrier.$dmhmap @(Writer w)
```